### PR TITLE
add computeAngleVertexNormals: calculation of normals taking into account an angle to neighboring polygons 

### PR DESCRIPTION
**Pull request by Ni55aN**
Example: https://codepen.io/Ni55aN/pen/zROmoe

### CODE REVIEW

File: src/core/Geometry.js [hunk 1]
```diff
@@ -545,6 +545,58 @@ Geometry.prototype = Object.assign( Object.create( EventDispatcher.prototype ),
 
 	},
 
+	computeAngleVertexNormals: function ( angle ) {
+
+		function weightedNormal( normals, vector ) {
+
+		  var normal = new THREE.Vector3();
+
+			for ( var i = 0; i < normals.length; i ++ ) {
```
Comment (Glinkis) [inline src/core/Geometry.js]: Reduce the indentation of this loop by one step.

**Comment by Ni55aN** [comment]
@Galiaf47 yes, sure

File: src/core/Geometry.js [hunk 2]
```diff
@@ -545,6 +545,58 @@ Geometry.prototype = Object.assign( Object.create( EventDispatcher.prototype ),
...
+
+			}
+
+		  return normal.normalize();
+
+		}
+
+		this.computeFaceNormals();
+
+		var vertexNormals = this.vertices.map( () => [] );
```
Comment (Mugen87) [inline src/core/Geometry.js]: Please avoid the usage of an array function.

File: src/core/Geometry.js [hunk 3]
```diff
@@ -545,6 +545,58 @@ Geometry.prototype = Object.assign( Object.create( EventDispatcher.prototype ),
...
+
+		for ( var i = 0; i < this.faces.length; i ++ ) {
```
[the remark this answers is not in the document]
Comment (Glinkis) [reply src/core/Geometry.js]: Why are the loops written like this?
In all modern browsers performance is pretty much the same if if you don't cache the length.

[the remark this answers is not in the document]
Comment (Mugen87) [reply src/core/Geometry.js]: > I understand if it's a stylistic choice

Um, I remember benchmarks that showed that this approach is a bit faster. Anyway, we are not using this style of `for` loop at all places in the code base. I've made this suggestion so the code of `computeAngleVertexNormals()` is more consistent to the other `compute*Normals()` methods.

File: src/core/Geometry.js (hunk 2 above)
Comment (Ni55aN) [reply src/core/Geometry.js]: hm.. three.js not welcome with es6?

**Comment by Ni55aN** [comment]
@Glinkis implementation for `BufferGeometry` looks more difficult (due to non-indexed geometry). I suggest to leave this for later, but for now we can use `fromBufferGeometry` / `fromGeometry`

**Comment by WestLangley** [comment]
This is a thing-of-beauty on extruded `TextGeometry`: smooths the faceted edges only. Likewise for `ExtrudeGeometry`.

screen shot 2018-09-06 at 6 28 36 pm

I'd prefer the following API, instead, however. It would be a significant improvement to the existing method, which is limited in this regard.
```javascript
computeVertexNormals( areaWeighted, creaseAngle ); // angle in radians; undefined same as Math.PI
```

It is unfortunate it is for `Geometry` only, not `BufferGeometry`, but I'll take it.

**Comment by Ni55aN** [comment]
@WestLangley the present `computeVertexNormals()` contains many of logic. I don't think it is good idea to add some code to this method. Or maybe it will only call the method if there is a second parameter
```js
computeVertexNormals( areaWeighted, creaseAngle ){
   if(creaseAngle) return computeAngleVertexNormals(creaseAngle)
}
```
but then the first parameter is lost

**Comment by WestLangley** [comment]
I think it would be OK to always use area-weighted:
```javascript
computeVertexNormals( creaseAngle ); // creaseAngle optional
```
(We know how to change the API in a backwards-compatible way.)

I do realize that @mrdoob does not want to make any more changes to `Geometry`, so this entire PR may be moot.

**Comment by Ni55aN** [comment]
Moving to 'utils' it would be good solution

**Comment by Glinkis** [comment]
My colleague has now written a utility function that does exactly this, but for `BufferGeometry`,
so should I issue a separate pull request to put this function in `BufferGeometryUtils` or something?

**Comment by Mugen87** [comment]
I vote for `BufferGeometry`.

**Comment by Glinkis** [comment]
Yes, me too. But I suppose it's ultimately up to @mrdoob ?

**Comment by WestLangley** [comment]
`BufferGeometry.computeVertexNormals( creaseAngle )` will only support `creaseAngle` for non-indexed. If indexed, users will have to convert their geometry to non-indexed first.

**Comment by WestLangley** [comment]
>Would it be all right to issue a separate pull request for inserting said function in either BufferGeometry, or BufferGeometryUtils?

Yes, but I expect it will require a restructuring of the data if the geometry is indexed. It may be better as a utility, in that case. I'd propose it as a utility at first. It can always be promoted.

**Comment by Glinkis** [comment]
Yes, it needs to split points for hard edges, and offset eventual geometry groups accordingly.

**Comment by Mugen87** [comment]
Closing, since `Geometry` will not be enhanced with such methods anymore. It's better to make a new PR with an enhancement for `BufferGeometryUtils`. I think this will have more chance to get merged...

**Pull request closed without merging**